src/main/java/com/google/api/codegen/transformer/csharp/CSharpSurfaceNamer.java

```diff
@@ -30,16 +30,78 @@
 import com.google.api.tools.framework.model.Interface;
 import com.google.api.tools.framework.model.Method;
 import com.google.api.tools.framework.model.TypeRef;
+import com.google.common.base.Function;
 import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
 import java.util.List;
+import java.util.Map;
 
 public class CSharpSurfaceNamer extends SurfaceNamer {
 
+  public static final class MethodFields {
+    public MethodFields(Method method, List<Field> fields) {
```

Comment:
Will AutoValue not work for this class? (I'm suspecting that using method.getFullName() for hashCode/equals would get in the way of that)

(entire class becomes obsolete with my suggestion about redoing generateMethods)